test/com/google/enterprise/adaptor/DocumentHandlerTest.java

```diff
@@ -402,6 +406,56 @@ public void getDocContent(Request request, Response response)
   }
 
   @Test
+  public void testContentTransform() throws Exception {
+    ContentTransformFactory contentTransformFactory =
+        new ContentTransformFactory(
+        new ArrayList<Map<String, String>>() {
+          {
+            add(new HashMap<String, String>() {
+              {
+                put("class", SampleDocumentContentTransform.class.getName());
+              }
+            });
+          }
+        });
+    mockAdaptor = new MockAdaptor() {
+      @Override
+      public void getDocContent(final Request request, final Response response)
+          throws IOException, InterruptedException {
+        response.setContentType("image/jpeg");
+        OutputStream os = response.getOutputStream();
+        os.write("some cool stuff".getBytes(Charsets.UTF_8));
+        os.close();
+      }
+    };
+    String remoteIp = ex.getRemoteAddress().getAddress().getHostAddress();
+    DocumentHandler handler = createHandlerBuilder()
+        .setAdaptor(mockAdaptor)
+        .setFullAccessHosts(new String[]{remoteIp})
+        .setContentTransformPipeline(contentTransformFactory)
+        .build();
+    handler.handle(ex);
+    assertEquals("some changed stuff", new String(ex.getResponseBytes()));
+  }
+
+  private static class SampleDocumentContentTransform
+      extends DocumentContentTransform {
+
+    public SampleDocumentContentTransform(Map<String, String> config) {
+      super(config);
+    }
+
+    @Override
+    public void write(final byte[] b) throws IOException {
+      if (contentType.equals("image/jpeg")) {
```

Comment:
Thank you. Added further comments.


- technology's compounding interest
-

On Mon, Aug 17, 2015 at 12:23 AM, Dominik Weidenfeld <
notifications@github.com> wrote:

> In test/com/google/enterprise/adaptor/DocumentHandlerTest.java
> <https://github.com/googlegsa/library/pull/7#discussion_r37163266>:
>
> > +        .setContentTransformPipeline(contentTransformFactory)
> > +        .build();
> > +    handler.handle(ex);
> > +    assertEquals("some changed stuff", new String(ex.getResponseBytes()));
> > +  }
> > +
> > +  private static class SampleDocumentContentTransform
> > +      extends DocumentContentTransform {
> > +
> > +    public SampleDocumentContentTransform(Map<String, String> config) {
> > +      super(config);
> > +    }
> > +
> > +    @Override
> > +    public void write(final byte[] b) throws IOException {
> > +      if (contentType.equals("image/jpeg")) {
>
> I have added the TODO, but we need to take care that the
> contentTransformer.getFinalContentType() (or similar) method is called
> before startSending().
> Maybe you want to watch the binary content first before changing the
> content-type... this would not be possible...
>
>     if ("HEAD".equals(ex.getRequestMethod())) {
>         // Unfortunately, we won't be able to report any errors after this
>         // point. We don't delay sending the headers, however, because of the
>         // watchdog.
>         state = State.HEAD;
>         startSending(false);
>         os = new SinkOutputStream();
>       } else {
>         state = State.SEND_BODY;
>         startSending(true);
>         countingOs = new CountingOutputStream(new CloseNotifyOutputStream(
>             ex.getResponseBody()));
>         os = countingOs;
>       }
>       if (null != contentTransformFactory) {
>         // TODO: possibility for changing the content-type in transformer
>         return contentTransformFactory
>             .createPipeline(os, contentType, metadata);
>       }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/googlegsa/library/pull/7/files#r37163266>.
>